managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java

```diff
@@ -2140,6 +2141,28 @@ void testIndividuallyDeletedMessages3(boolean useOpenRangeSet) throws Exception
         assertTrue(c1.isIndividuallyDeletedEntriesEmpty());
     }
 
+    @Test(timeOut = 20000)
+    void testFindNewestMatchingAfterLedgerRollover() throws Exception {
+        ManagedLedgerImpl ledger = (ManagedLedgerImpl) factory.open("my_test_ledger");
+        ManagedCursorImpl c1 = (ManagedCursorImpl) ledger.openCursor("c1");
+        ledger.addEntry("expired".getBytes(Encoding));
+        ledger.addEntry("expired".getBytes(Encoding));
+        ledger.addEntry("expired".getBytes(Encoding));
+        ledger.addEntry("expired".getBytes(Encoding));
+        Position last = ledger.addEntry("expired".getBytes(Encoding));
+
+        // roll a new ledger
+        int numLedgersBefore = ledger.getLedgersInfo().size();
+        ledger.getConfig().setMaxEntriesPerLedger(1);
+        ledger.rollCurrentLedgerIfFull();
+        Awaitility.await().atMost(20, TimeUnit.SECONDS)
+                .until(() -> ledger.getLedgersInfo().size() > numLedgersBefore);
+
+        assertEquals(last,
+                c1.findNewestMatching(entry -> Arrays.equals(entry.getDataAndRelease(), "expired".getBytes(Encoding))));
```

Comment:
This test doesn't make any sense to me. You are adding 5 entries with the same content.

Reply:
it is the same thing that happens with "expired" messages.

If you have all of the messages "expired" the condition evaluates to true for every message ("message is expired"), and in that case findNewestMatching must return the last message in the topic.

Reply:
But you can't tell if "expired" message is the last message. Because all the messages have the same content.

Reply:
I see your point, let me update the test. thanks for the clarification